payment/app.py

```diff
@@ -12,39 +14,164 @@
+        new_user_id = f"user:{random.getrandbits(ID_BYTES_SIZE)}"
+    
+    new_user = User(user_id=new_user_id)
+    
+    try:
+        db.hset(new_user.user_id, mapping=new_user.to_dict())
+    except Exception as exp:
+        return Response(str(exp), status=400)
+
+    return Response(json.dumps(new_user.to_dict()), mimetype="application/json", status=200)
 
 
 @app.get('/find_user/<user_id>')
 def find_user(user_id: str):
-    pass
+    
+    user = db.hgetall(user_id)
+
+    if not user:
+        return Response(f"The user {user_id} does not exist in the DB!", status=404)
+    
+    return_item = {
+        "user_id" : user[b"user_id"].decode("utf-8")
+        "credit" : int(user[b"credit"])
+    }
+    return Response(return_item, mimetype="application/json", status=200)
 
 
 @app.post('/add_funds/<user_id>/<amount>')
 def add_credit(user_id: str, amount: int):
-    pass
-
+    
+    if int(amount) <= 0:
+        body = {
+            "done" : False
+        }
+        return Response(json.dumps(body), mimetype="application/json", status=200)
+
+    if not db.hget(user_id, "user_id"):
+        body = {
+            "done" : False
+        }
+        return Response(json.dumps(body), mimetype="application/json", status=200)
+
+    try:
+        new_amount = db.hincrby(user_id, "credit", int(amount))
+    except Exception as err:
+        return Response(str(err), status=400)
+
+    body = {
+        "done" : True
+    }
+
+    return Response(json.dumps(body), mimetype="application/json", status=200)
 
 @app.post('/pay/<user_id>/<order_id>/<amount>')
 def remove_credit(user_id: str, order_id: str, amount: int):
-    pass
+    
+    if int(amount) <= 0:
+        return Response("The amount must be >0!", status=400)
+
+    if not db.hget(user_id, "user_id"):
+        return Response(f"The user {user_id} does not exist in the DB!", status=404)
+
+    order_find_path = gateway_url + f"order-service/orders/find/{order_id}"
+    r = requests.get(order_find_path)
+    response_json = json.load(r.json())
+
+    if r.status_code==404:
+        return Response(f"The order {order_id} does not exist in the DB!", status=404)
 
+    if bool(response_json["paid"]):
+        return Response(f"The order {order_id} has been already paid", status=400)
+
+    current_credit = int(db.hget(user_id, "credit"))
+
+    if current_credit < int(amount):
+        return Response(f"Insufficient credit balance")
+
+    try:
+        new_credit = db.hincrby(user_id, "credit", -1*int(amount))
+    except Exception as err:
+        return Response(str(err), status=400)
+
+    return Response(f"The payment of the order {order_id} is paid", status=200)
 
 @app.post('/cancel/<user_id>/<order_id>')
 def cancel_payment(user_id: str, order_id: str):
-    pass
-
+    if not db.hget(user_id, "user_id"):
+        return Response(f"The user {user_id} does not exist in the DB!", status=404)
+    
+    order_find_path = gateway_url + f"order-service/orders/find/{order_id}"
```

Comment:
Is this URL working correctly?

Reply:
These urls should be fixed to the correct ones but it can only be tested after merging.